### PR TITLE
master:fix-previous-value

### PR DESCRIPTION
**Pull request by nemisj**
### Pull Request check-list

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?

### Description of change

Previous method of the instance was working incorrectly

### CODE REVIEW

**Comment by nemisj** [comment]
@sushantdhiman , added

File: test/unit/instance/previous.test.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+/* jshint -W030 */
+var chai = require('chai')
+  , expect = chai.expect
+  , Support   = require(__dirname + '/../support')
+  , DataTypes = require(__dirname + '/../../../lib/data-types')
+  , current   = Support.sequelize;
+
+describe(Support.getTestDialectTeaser('Instance'), function() {
```
Comment (sushantdhiman) [inline test/unit/instance/previous.test.js]: Space before opening brackets `function ()`

**Comment by nemisj** [comment]
done

**Comment by BorisKozo** [comment]
This fix broke our application :D
Reminded me of this: https://xkcd.com/1172/

**Comment by barriber** [comment]
Previous function doesn't return the correct values now,

```
return Promise.map(instances, function(instance) {
          // Record updates in instances dataValues
          Utils._.extend(instance.dataValues, values);
          // Set the changed fields on the instance
          Utils._.forIn(valuesUse, function(newValue, attr) {
            if (newValue !== instance._previousDataValues[attr]) {
              instance.setDataValue(attr, newValue);
            }
          });

```
instance.dataValues override with the new values in  Utils._.extend(instance.dataValues, values);
and  in instance.setDataValue(attr, newValue); 


```
Instance.prototype.setDataValue = function(key, value) {
  var originalValue = this.dataValues[key];
  if (!Utils.isPrimitive(value) || value !== originalValue) {
    this.changed(key, true);
  }

  this._previousDataValues[key] = originalValue;
  this.dataValues[key] = value;
};
```

the _previousDataValues overrided with same new data value

**Comment by nemisj** [comment]
@barriber so actually we also have to change that code which sets dataValues directly I guess, into something like:

```
return Promise.map(instances, function(instance) {
          // Set the changed fields on the instance
          Utils._.forIn(valuesUse, function(newValue, attr) {
              instance.setDataValue(attr, newValue);
          });
```

in order to guarantee the same data flow into the instance

**Comment by barriber** [comment]
@janmeier could you please comment if this is the correct solution?

**Comment by nemisj** [comment]
@barriber i see that there is a note in `changelog.md` saying that from 2.0.0. version _previousDataValues must be updated after `afterUpdate`. But I don't understand this comment completely why it's like that. So i guess this solution is also not good and should be done differently.

File: lib/model.js [hunk 2]
```diff
@@ -2977,11 +2977,12 @@ class Model {
    * @param {any} value
    */
   setDataValue(key, value) {
```
Comment (Verdier) [inline lib/model.js]: This does't work if you call setDataValue multiple times... `previous` should return the original value (ie. the value in the DB when the record was fetched), am I wrong?

Comment (Verdier) [inline lib/model.js]: Ok that was reverted... sorry ^^

**Pull request merged**